libs/server/Custom/ExpandableMap.cs

```diff
@@ -0,0 +1,334 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Threading;
+
+namespace Garnet.server
+{
+    /// <summary>
+    /// This class defines a map of items of type T whose keys are a specified range of IDs (can be descending / ascending)
+    /// The size of the underlying array containing the items doubles in size as needed
+    /// </summary>
+    /// <typeparam name="T">Type of item to store</typeparam>
+    internal struct ExpandableMap<T>
+    {
+        /// <summary>
+        /// The underlying array containing the items
+        /// </summary>
+        internal T[] map;
+
+        /// <summary>
+        /// Reader-writer lock for the underlying item array
+        /// </summary>
+        internal readonly ReaderWriterLockSlim mapLock = new();
```

Comment:
Can we use @vazois's `SingleWriterMultiReaderLock`?

Reply:
I don't think it supports upgrading an acquired read lock to a write lock, right @vazois?

Reply:
why does it need upgradability? this might be more efficient than what C# provides.

Reply:
Thought about it, it's fine as is :)